sdk/modelsrepository/Azure.Iot.ModelsRepository/src/RepositoryHandler.cs

```diff
@@ -0,0 +1,118 @@
+        {
+            Dictionary<string, string> processedModels = new Dictionary<string, string>();
+            Queue<string> toProcessModels = new Queue<string>();
+
+            foreach (string dtmi in dtmis)
+            {
+                if (!DtmiConventions.IsDtmi(dtmi))
+                {
+                    ResolverEventSource.Shared.InvalidDtmiInput(dtmi);
+                    string invalidArgMsg = string.Format(CultureInfo.InvariantCulture, StandardStrings.InvalidDtmiFormat, dtmi);
+                    throw new ResolverException(dtmi, invalidArgMsg, new ArgumentException(invalidArgMsg));
+                }
+
+                toProcessModels.Enqueue(dtmi);
+            }
+
+            while (toProcessModels.Count != 0 && !cancellationToken.IsCancellationRequested)
+            {
+                string targetDtmi = toProcessModels.Dequeue();
+                if (processedModels.ContainsKey(targetDtmi))
+                {
+                    ResolverEventSource.Shared.SkippingPreprocessedDtmi(targetDtmi);
+                    continue;
+                }
+                ResolverEventSource.Shared.ProcessingDtmi(targetDtmi);
+
+                FetchResult result = await FetchAsync(targetDtmi, cancellationToken).ConfigureAwait(false);
+                if (result.FromExpanded)
+                {
+                    Dictionary<string, string> expanded = await new ModelQuery(result.Definition).ListToDictAsync().ConfigureAwait(false);
+                    foreach (KeyValuePair<string, string> kvp in expanded)
+                    {
+                        if (!processedModels.ContainsKey(kvp.Key))
+                            processedModels.Add(kvp.Key, kvp.Value);
+                    }
+
+                    continue;
+                }
+
+                ModelMetadata metadata = new ModelQuery(result.Definition).GetMetadata();
+
+                if (ClientOptions.DependencyResolution >= DependencyResolutionOption.Enabled)
+                {
+                    IList<string> dependencies = metadata.Dependencies;
+
+                    if (dependencies.Count > 0)
+                        ResolverEventSource.Shared.DiscoveredDependencies(string.Join("\", \"", dependencies));
+
+                    foreach (string dep in dependencies)
+                    {
+                        toProcessModels.Enqueue(dep);
+                    }
+                }
+
+                string parsedDtmi = metadata.Id;
+                if (!parsedDtmi.Equals(targetDtmi, StringComparison.Ordinal))
+                {
+                    ResolverEventSource.Shared.IncorrectDtmiCasing(targetDtmi, parsedDtmi);
+                    string formatErrorMsg = string.Format(CultureInfo.InvariantCulture, StandardStrings.IncorrectDtmiCasing, targetDtmi, parsedDtmi);
+                    throw new ResolverException(targetDtmi, formatErrorMsg, new FormatException(formatErrorMsg));
```

Comment:
We'll need to not throw custom exception types, and instead use RequestFailedException from Azure.Core.

Reply:
I do have a task to go over these